File: 503-next-greater-element-ii/next-greater-element-ii.cpp
#include <stack>
#include <vector>
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
        std::stack<int> st;
        std::vector<int> res;
        int n = nums.size();
        int j = 0;
        for (int i = 2*n-1; i >=0; i--)   {
            while (!st.empty() and st.top() <= nums[i%n])
                st.pop();
            if (i < n)  {
                int r = st.empty()? -1 : st.top();
                res.push_back(r);
            }
            st.push(nums[i%n]);
        }
        res = vector<int>(res.rbegin(), res.rend());
        return res;
    }
};

File: 503-next-greater-element-ii/next-greater-element-ii.py
from collections import deque
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        res = []
        n = len(nums)
        i = 2*n-1
        st = deque()
        while (i >= 0):
            while (st and st[-1] <= nums[i%n]):
                st.pop()
            if i < n:
                res.append(st[-1] if st else -1)
            st.append(nums[i%n])
            i -= 1
        return res[::-1]
        